encoding/codecv2/codecv2.go

```diff
@@ -0,0 +1,618 @@
+
+// Encode serializes the DABlock into a slice of bytes.
+func (b *DABlock) Encode() []byte {
+	bytes := make([]byte, 60)
+	binary.BigEndian.PutUint64(bytes[0:], b.BlockNumber)
+	binary.BigEndian.PutUint64(bytes[8:], b.Timestamp)
+	if b.BaseFee != nil {
+		binary.BigEndian.PutUint64(bytes[40:], b.BaseFee.Uint64())
+	}
+	binary.BigEndian.PutUint64(bytes[48:], b.GasLimit)
+	binary.BigEndian.PutUint16(bytes[56:], b.NumTransactions)
+	binary.BigEndian.PutUint16(bytes[58:], b.NumL1Messages)
+	return bytes
+}
+
+// DecodeDABlock takes a byte slice and decodes it into a DABlock.
+func DecodeDABlock(bytes []byte) (*DABlock, error) {
+	if len(bytes) != 60 {
+		return nil, errors.New("block encoding is not 60 bytes long")
+	}
+
+	block := &DABlock{
+		BlockNumber:     binary.BigEndian.Uint64(bytes[0:8]),
+		Timestamp:       binary.BigEndian.Uint64(bytes[8:16]),
+		BaseFee:         new(big.Int).SetUint64(binary.BigEndian.Uint64(bytes[40:48])),
+		GasLimit:        binary.BigEndian.Uint64(bytes[48:56]),
+		NumTransactions: binary.BigEndian.Uint16(bytes[56:58]),
+		NumL1Messages:   binary.BigEndian.Uint16(bytes[58:60]),
+	}
+
+	return block, nil
+}
+
+// NewDAChunk creates a new DAChunk from the given encoding.Chunk and the total number of L1 messages popped before.
+func NewDAChunk(chunk *encoding.Chunk, totalL1MessagePoppedBefore uint64) (*DAChunk, error) {
+	var blocks []*DABlock
+	var txs [][]*types.TransactionData
+
+	for _, block := range chunk.Blocks {
+		b, err := NewDABlock(block, totalL1MessagePoppedBefore)
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, b)
+		totalL1MessagePoppedBefore += block.NumL1Messages(totalL1MessagePoppedBefore)
+		txs = append(txs, block.Transactions)
+	}
+
+	daChunk := DAChunk{
+		Blocks:       blocks,
+		Transactions: txs,
+	}
+
+	return &daChunk, nil
+}
+
+// Encode serializes the DAChunk into a slice of bytes.
+func (c *DAChunk) Encode() []byte {
+	var chunkBytes []byte
+	chunkBytes = append(chunkBytes, byte(len(c.Blocks)))
+
+	for _, block := range c.Blocks {
+		blockBytes := block.Encode()
+		chunkBytes = append(chunkBytes, blockBytes...)
+	}
+
+	return chunkBytes
+}
+
+// Hash computes the hash of the DAChunk data.
+func (c *DAChunk) Hash() (common.Hash, error) {
+	var dataBytes []byte
+
+	// concatenate block contexts
+	for _, block := range c.Blocks {
+		encodedBlock := block.Encode()
+		// only the first 58 bytes are used in the hashing process
+		dataBytes = append(dataBytes, encodedBlock[:58]...)
+	}
+
+	// concatenate l1 tx hashes
+	for _, blockTxs := range c.Transactions {
+		for _, txData := range blockTxs {
+			if txData.Type == types.L1MessageTxType {
+				txHash := strings.TrimPrefix(txData.TxHash, "0x")
+				hashBytes, err := hex.DecodeString(txHash)
+				if err != nil {
+					return common.Hash{}, err
+				}
+				if len(hashBytes) != 32 {
+					return common.Hash{}, fmt.Errorf("unexpected hash: %s", txData.TxHash)
+				}
+				dataBytes = append(dataBytes, hashBytes...)
+			}
+		}
```

Comment:
```suggestion
		for _, txData := range blockTxs {
		        if txData.Type != types.L1MessageTxType {
		             continue
		        }
		        
			txHash := strings.TrimPrefix(txData.TxHash, "0x")
			hashBytes, err := hex.DecodeString(txHash)
			if err != nil {
				return common.Hash{}, err
			}
			if len(hashBytes) != 32 {
				return common.Hash{}, fmt.Errorf("unexpected hash: %s", txData.TxHash)
			}
			dataBytes = append(dataBytes, hashBytes...)
		}
```

Reply:
refactored in [67cc822](https://github.com/scroll-tech/da-codec/pull/2/commits/67cc822e9db69b8d112f610b65167e3501729ba3).